pkg/controller/volume/vacprotection/vac_protection_controller_test.go

```diff
@@ -0,0 +1,345 @@
+			},
+			expectedActions: []clienttesting.Action{
+				// This fails
+				clienttesting.NewUpdateAction(vacGVR, "", vac1WithFinalizer),
+				// This fails too
+				clienttesting.NewUpdateAction(vacGVR, "", vac1WithFinalizer),
+				// This succeeds
+				clienttesting.NewUpdateAction(vacGVR, "", vac1WithFinalizer),
+			},
+		},
+		{
+			name:       "deleted VAC with finalizer -> finalizer is removed",
+			updatedVAC: vac1TerminatingWithFinalizer,
+			expectedActions: []clienttesting.Action{
+				clienttesting.NewUpdateAction(vacGVR, "", vac1Terminating),
+			},
+		},
+		{
+			name:       "finalizer removal fails -> controller retries",
+			updatedVAC: vac1TerminatingWithFinalizer,
+			reactors: []reaction{
+				{
+					verb:      "update",
+					resource:  "volumeattributesclasses",
+					reactorfn: generateUpdateErrorFunc(t, 2 /* update fails twice*/),
+				},
+			},
+			expectedActions: []clienttesting.Action{
+				// Fails
+				clienttesting.NewUpdateAction(vacGVR, "", vac1Terminating),
+				// Fails too
+				clienttesting.NewUpdateAction(vacGVR, "", vac1Terminating),
+				// Succeeds
+				clienttesting.NewUpdateAction(vacGVR, "", vac1Terminating),
+			},
+		},
+		{
+			name:            "deleted VAC with finalizer but it's referenced by a PV -> finalizer is not removed",
+			initialObjects:  []runtime.Object{pv1WithVAC1},
+			updatedVAC:      vac1TerminatingWithFinalizer,
+			expectedActions: []clienttesting.Action{},
+		},
+		{
+			name:            "deleted VAC with finalizer but it's referenced by a PVC -> finalizer is not removed",
+			initialObjects:  []runtime.Object{pvc1WithVAC1},
+			updatedVAC:      vac1TerminatingWithFinalizer,
+			expectedActions: []clienttesting.Action{},
+		},
+		// PV events
+		//
+		{
+			name:           "pv changes vac and deleted VAC with finalizer -> finalizer is removed",
+			initialObjects: []runtime.Object{vac1TerminatingWithFinalizer},
+			oldPV:          pv1WithVAC1,
+			updatedPV:      pv1WithVAC2,
+			expectedActions: []clienttesting.Action{
+				clienttesting.NewUpdateAction(vacGVR, "", vac1Terminating),
+			},
+		},
+		{
+			name:           "pv is deleted and deleted VAC with finalizer -> finalizer is removed",
+			initialObjects: []runtime.Object{vac1TerminatingWithFinalizer},
+			oldPV:          pv1WithVAC1,
+			expectedActions: []clienttesting.Action{
+				clienttesting.NewUpdateAction(vacGVR, "", vac1Terminating),
+			},
+		},
+		{
+			name:            "pv is deleted but other pv still holds terminating vac -> finalizer is not removed",
+			initialObjects:  []runtime.Object{vac1TerminatingWithFinalizer, pv2WithVAC1},
+			oldPV:           pv1WithVAC1,
+			expectedActions: []clienttesting.Action{},
+		},
+		{
+			name:            "pv is deleted but pvc still holds terminating vac -> finalizer is not removed",
+			initialObjects:  []runtime.Object{vac1TerminatingWithFinalizer, pvc1WithVAC1},
+			oldPV:           pv1WithVAC1,
+			expectedActions: []clienttesting.Action{},
+		},
+		// PVC events
```

Comment:
If the targetVolumeAttributesClass of a PVC's modifyVolumeStatus is "vac2", can "vac2" be deleted from kube-apiserver?

Reply:
We should not delete targetVolumeAttributesClass because it is something the pvc is trying to update to.

Reply:
a unit test is added for this case.